src/effcheck.ml

```diff
@@ -82,16 +86,32 @@ let rec tcheck env term =
         lst;
       (typ, eff)
     | _ -> Test.fail_report "tcheck: ListTrm must have a list type")
-  | Constructor (typ, name, payload_lst, eff) ->
-    (match check_opt_invariants (typ, name, payload_lst) with
-    | Ok _ ->
-      List.iter
-        (fun trm ->
-          (* ignore value but propatage error *)
-          tcheck env trm |> ignore)
-        payload_lst;
-      (typ, eff)
-    | Error msg -> Test.fail_report msg)
+  | Constructor (typ, descr, _, eff) ->
```

Comment:
You could have three cases

```ocaml
| Constructor (Option ty, Variant variant, args, eff) ->
  ...
| Constructor (List tys, TupleArity arity, args, eff) ->
   ...
| Constructor (typ, variant, args, eff) ->
   error message
```

I think it would be nice to factorize the second part in a `check_list_invars tys arity args eff` function, to keep the main function readable.

Reply:
I want to keep two cases for Constructors by `constr_descr`: variants and tuples to make checking invariants for those terms easier and to have more specific error reporting.

Reply:
Having specific error messages can be done by having three cases: one for options, one for lists, and one (always an error) for unknown types. But my intuition is that it's more natural to match on the type first, and then on the constructors, instead of the other way around as you do. Reasons include:

- Readers of the code may not know the details of our representation of constructors (what `TupleArity` means), but they know the types of the language, so checking on that part first clarifies the code for them.
- It is the knowledge of the type that tells us what constructors should be valid for it, rather than the other way around (for example, in OCaml it is now valid to have distinct types share a variant constructor of the same name, so knowing the constructor doesn't tell you which type to check, it's the other way around).
- In particular, one could imagine a OCaml-like language where a specific algebraic datatype has a mix of Tuple-style constructors and also normal constructors; in that case the strategy of first matching on the constructor clearly does not work very well. (This is hypothetical as OCaml does not allow this; but for example if we decided to handle `(::)` and `[]` as another case of constructors (than `Variant` and `TupleArity`, maybe `Infix` or something), then mixing those and normal constructors in a datatype would be valid.)

Reply:
That is convincing. I will work on this after adding elim rules.